drift.py

```diff
@@ -1,146 +1,115 @@
-import argparse
-import json, requests
-import subprocess, sys
+from spack.cmd.diff import compare_specs
+from spack.spec import Spec
+from spack.version import ver
 
 # Define SUCCESS for comparing command return codes.
 SUCCESS = 0
-args = None
 
+@dataclass
 class Result:
-    abstract:str
+    name:str
+    version:str
     commit:str
     tags:list
     timestamp:str
-    spec:str
-    version:str
-
-    def __init__(self):
-        self.abstract = ""
-        self.commit = ""
-        self.tags = []
-        self.timestamp = ""
-        self.spec = ""
-        self.version = ""
-
 
 class DriftAnalysis(Helicase):
-    specs = []
-    last = {}
-    def __init__(self, specs):
-        self.specs = specs
-        for spec in specs:
-            self.last[spec] = ""
+    def __init__(self, specs=None):
+        """Setup the last cache and specs used for the analysis."""
+
+        self.last = {}
+        self.specs = specs or []
 
     def analyze(self, commit):
-        for spec in self.specs:
-            out = spack("spec --json " + spec)
-            verToken = spec.find("@")
-            if verToken < 0: verToken = len(spec)
-            name = spec[:verToken]
-            result = Result()
-            result.commit = commit.hash
-            result.abstract = name
-            result.timestamp = str(commit.author_date)
-            if len(spec[verToken:]) > 0:
-                result.version = spec[verToken+1:]
-            if out.returncode == SUCCESS:
-                concrete_spec = json.loads(out.stdout)["spec"]
-                # In the case that the concretized spec hash changes
-                # we want to record the commit where this happens.
-                if self.last[spec] != "" and self.last[spec][0][name]["full_hash"] != concrete_spec[0][name]["full_hash"]:
-                    # Check if the package version is different.
-                    if self.last[spec][0][name]["version"] != concrete_spec[0][name]["version"]:
-                        result.tags += ["pkg-updated"]
-                    # +---------------------+
-                    # |     Dependencies    |
-                    # +---------------------+
-                    result.tags += set(self.tag_deps(concrete_spec, spec, name))
-                    # +---------------------+
-                    # |      Variants       |
-                    # +---------------------+
-                    result.tags += set(self.tag_variants(concrete_spec, spec, name))
-                    # Send Inflection Point
-                    send(result)
-                # Save concrete spec hash as last hash.
-                self.last[spec] = concrete_spec
-            else:
-                # If the spec doesn't concretize properly we also want
-                # to record the commit at which this occurred.
-                result.tags += ["concretizaton-failed"]
-                send(result)
+        """Analyze defines the process that should be run on every commit
```

Comment:
If you want to follow Todd documentation convention (e.g., spack) what we do is have the first line be a single line to describe the function, and then an empty line, and then more verbose description below. But you don't have to, as this is not spack.

Reply:
It's from here: https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

Reply:
it is forever ingrained in my mind as "the Todd documentation convention," heh.